h2o-algos/src/main/java/hex/tree/GuidedSplitPoints.java

```diff
@@ -0,0 +1,159 @@
+package hex.tree;
+
+import water.util.ArrayUtils;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Implements a method for finding new histogram bins split-points based on a result of previous binning.
+ * Idea:
+ *   We take non-empty bins and look at the squared error they have. Based on the target bin number, we discard
+ *   the empty bins and used the frees-up space to refine the non-non empty bins. Splitting of non-empty bins
+ *   is guided by Squared Error accumulated in the bin. Bins with higher SE are split more than the bins with lower SE.
+ *   Sub-bins (bins created from a single original bin) are refined uniformly.
+ *   
+ *   If uniform splitting fails in this iteration (= the distribution of values is significantly skewed), next iteration
+ *   will attempt correct the issue by repeating the procedure with new bins (we are recursively refining the promising
+ *   bins as we get deeper in the tree).
+ */
+public class GuidedSplitPoints {
+
+    static final double LOW_DENSITY_THRESHOLD = 0.2;
+
+    static boolean isApplicableTo(DHistogram h) {
+        return h._vals != null &&  // observations were not yet binned, we don't have the data to guide the splitting
+                h._isInt != 2 &&   // categorical columns have a specific handing
+                !h._intOpt;        // integer optimized columns have a single value per bin, no point in refining such bins
+    }
+
+    static double[] makeSplitPoints(DHistogram h, final int targetNBins, final double min, final double maxEx) {
+        // Collect bins to consider for refining
+        final List<BinDescriptor> bins = extractNonEmptyBins(h);
+
+        // Budget is given by target number of bins in the new layer, we keep all non-empty bins
+        final int totalBudget = targetNBins - bins.size() - 2; // how many bins we have to allocate (save 2 spots for min/max)
```

Comment:
What if min or max are already in bins?

Reply:
That can happen, there will be eliminated by the last step:
```
return ArrayUtils.makeUniqueAndLimitToRange(customSplitPoints, min, maxEx);
```
It is problematic to handle this exactly right because of rounding errors - I prefer not to worry about it and let method makeUniqueAndLimit to range take care of it.

Reply:
Ok, sounds good.